[VectorUtils] Trivially vectorize ldexp, [l]lround

artagnon opened the pull request.

artagnon committed:
[VectorUtils] Trivially vectorize ldexp, [l]lround

fhahn reviewed (commented):
That should be fine, assuming the cost model returns reasonable costs for them. Would be good to check if we have tests for their cost (possibly just as cost-model tests)

artagnon commented:
> > That should be fine, assuming the cost model returns reasonable costs for them. Would be good to check if we have tests for their cost (possibly just as cost-model tests)
> 
> Thanks, I've added CostModel tests, although the scalable versions have invalid costs?

Scalable-vector versions cannot be lowered, so the invalid costs are correct. I've fixed the costs of [l]lround in https://github.com/llvm/llvm-project/pull/145714.

fhahn reviewed (commented):
I gave this a try on AArch64 and it seems to enable vectorizing a few more loops, but in some cases it doesn't seem profitable, one example being https://llvm.godbolt.org/z/nb7667c71

That's even though the cost of `ldexp` on AArch64 is already quite high: https://llvm.godbolt.org/z/Wb1nj3Knq

cc @davemgreen in case he has additional thoughts

fhahn commented on llvm/test/Analysis/CostModel/RISCV/exp.ll:
```diff
@@ -0,0 +1,109 @@
+; NOTE: Assertions have been autogenerated by utils/update_analyze_test_checks.py
```
Thanks for adding the extra tests, I think you can commit the tests separately.

davemgreen commented:
I would expect the costs to be something like 10 for a libcall, and 10*4+scalarization_overhead for the vector. They might need to be overriden on AArch64 as they are marked custom, but I don't see them in BasicTTIImplBase::getTypeBasedIntrinsicInstrCost.

Can you add tests for SLP / scalarizer, as in https://reviews.llvm.org/D124358 or https://github.com/llvm/llvm-project/pull/71416/files?

artagnon committed:
[SLPVectorizer/Scalarizer] Add tests for [l]lround, ldexp

arsenm reviewed (commented):
Can you add an SLP vectorizer test for ldexp, particularly one where they use different types for the exponent

artagnon committed:
[SLPVectorize/test] More extensive ldexp testing

artagnon commented:
Gentle ping.

omjavaid commented:
This broke various LLVM buildbots testing AArch64 SVE,  but the problem was masked because relevant buildbots were already failing due to another breakage.
I am going to revert this as it has been while since bots were red. Kindly look into the failure and push it again after the fix.
https://lab.llvm.org/buildbot/#/builders/4/builds/8164
https://lab.llvm.org/buildbot/#/builders/17/builds/9858
https://lab.llvm.org/buildbot/#/builders/41/builds/8067
https://lab.llvm.org/buildbot/#/builders/143/builds/9607

artagnon commented:
Thanks, the problem was indeed masked. I'll look into the issue.

davemgreen commented:
I see. The hoisted vplan node is not included in the cost calculation, even though it is vectorized.

artagnon commented:
> I see. The hoisted vplan node is not included in the cost calculation, even though it is vectorized.

@davemgreen Thanks for the investigation: this seems to be a bug in LV. Could you kindly share a LLVM IR reproducer?

davemgreen commented:
Does this fail for you? https://godbolt.org/z/17Ye168cK
The original is something like `flang llvm-test-suite/Fortran/gfortran/regression/vect/pr106253.f -O3 -march=armv9-a`

artagnon commented:
I tried running your example with `opt -passes='default<O3>' -mtriple=aarch64 -mattr=+sve`, but I get a scalar `tail call i32 @llvm.lround.i32.f32(float 0.000000e+00)` hoisted.

davemgreen commented:
Does it fail with just '-passes=loop-vectorize`? As in does it generate a `%14 = call <vscale x 4 x i32> @llvm.lround.nxv4i32.nxv4f32(<vscale x 4 x float> zeroinitializer)`?
